db/python/tables/bq/billing_base.py

```diff
@@ -0,0 +1,673 @@
+from collections import Counter, defaultdict, namedtuple
+from datetime import datetime
+from typing import Any
+
+from google.cloud import bigquery
+
+from api.settings import BQ_BUDGET_VIEW, BQ_DAYS_BACK_OPTIMAL
+from api.utils.dates import get_invoice_month_range, reformat_datetime
+from db.python.gcp_connect import BqDbBase
+from models.models import (
+    BillingColumn,
+    BillingCostBudgetRecord,
+    BillingTimePeriods,
+    BillingTotalCostQueryModel,
+)
+
+# Label added to each Billing Big Query request,
+# so we can track the cost of metamist-api BQ usage
+BQ_LABELS = {'source': 'metamist-api'}
+
+
+# Day Time details used in grouping and parsing formulas
+TimeGroupingDetails = namedtuple('TimeGroupingDetails', ['field', 'formula'])
+
+
+def abbrev_cost_category(cost_category: str) -> str:
+    """abbreviate cost category"""
+    return 'S' if cost_category == 'Cloud Storage' else 'C'
+
+
+def prepare_time_periods(
+    query: BillingTotalCostQueryModel,
+) -> TimeGroupingDetails:
+    """Prepare Time periods grouping and parsing formulas"""
+    time_column = query.time_column or 'day'
+    result = TimeGroupingDetails('', '')
+
+    # Based on specified time period, add the corresponding column
+    if query.time_periods == BillingTimePeriods.DAY:
+        result = TimeGroupingDetails(
+            field=f'FORMAT_DATE("%Y-%m-%d", {time_column}) as day, ',
+            formula='PARSE_DATE("%Y-%m-%d", day) as day, ',
+        )
+    elif query.time_periods == BillingTimePeriods.WEEK:
+        result = TimeGroupingDetails(
+            field=f'FORMAT_DATE("%Y%W", {time_column}) as day, ',
+            formula='PARSE_DATE("%Y%W", day) as day, ',
+        )
+    elif query.time_periods == BillingTimePeriods.MONTH:
+        result = TimeGroupingDetails(
+            field=f'FORMAT_DATE("%Y%m", {time_column}) as day, ',
+            formula='PARSE_DATE("%Y%m", day) as day, ',
+        )
+    elif query.time_periods == BillingTimePeriods.INVOICE_MONTH:
+        result = TimeGroupingDetails(
+            field='invoice_month as day, ', formula='PARSE_DATE("%Y%m", day) as day, '
+        )
+
+    return result
+
+
+def construct_filter(
+    name: str, value: Any, is_label: bool = False
+) -> tuple[str, bigquery.ScalarQueryParameter | bigquery.ArrayQueryParameter]:
+    """Based on Filter value, construct filter string and query parameter
+
+    Args:
+        name (str): Filter name
+        value (Any): Filter value
+        is_label (bool, optional): Is filter a label?. Defaults to False.
+
+    Returns:
+        tuple[str, bigquery.ScalarQueryParameter | bigquery.ArrayQueryParameter]
+    """
+    compare = '='
+    b1, b2 = '', ''
+    param_type = bigquery.ScalarQueryParameter
+    key = name.replace('-', '_')
+
+    if isinstance(value, list):
+        compare = 'IN'
+        b1, b2 = 'UNNEST(', ')'
+        param_type = bigquery.ArrayQueryParameter
+
+    if is_label:
+        name = f'getLabelValue(labels, "{name}")'
```

Comment:
Can we use the python quoting function here, or replace any `"` with an escaped value?
Or can these values be passed so bigquery fills them in?

Reply:
Let me try the bigquery fill, it would make more sense do it that way.

Reply:
This has been refactored, managed to wrapped this inside FunctionBQFilter:
https://github.com/populationgenomics/metamist/blob/14f29ffa9b6d4021ac507c0a29fba735a3a28736/db/python/tables/bq/function_bq_filter.py#L10